Introduce New Classes to Existing Codebase

stino-x opened the pull request:
This pull request introduces two new classes, Options and App, to the existing codebase of the School Library App. These new classes enhance the functionality of the application and improve its organization. The key changes in this pull request are as follows:

Class Introduction: The Options class provides a menu-driven interface for users to interact with the library app. The App class manages the library's books, people (students and teachers), and rentals. These classes are designed to enhance the organization and functionality of the application.

Menu System: The Options class introduces a menu system that allows users to choose various options, including listing books, listing people, creating people (students or teachers), creating books, creating rentals, and listing rentals for a given person ID.

Modular Code: The introduction of these classes results in a more modular and organized code structure, making it easier to maintain and expand the application in the future.


.

stino-x committed:
Introduce New Classes to Existing Codebase

This pull request introduces two new classes, Options and App, to the existing codebase of the School Library App. These new classes enhance the functionality of the application and improve its organization. The key changes in this pull request are as follows:

Class Introduction: The Options class provides a menu-driven interface for users to interact with the library app. The App class manages the library's books, people (students and teachers), and rentals. These classes are designed to enhance the organization and functionality of the application.

Menu System: The Options class introduces a menu system that allows users to choose various options, including listing books, listing people, creating people (students or teachers), creating books, creating rentals, and listing rentals for a given person ID.

Modular Code: The introduction of these classes results in a more modular and organized code structure, making it easier to maintain and expand the application in the future.

.

prolajumokeoni commented on app.rb:
```diff
@@ -0,0 +1,116 @@
+    @parent.show_menu
+  end
+
+  def create_person
+    puts 'Do you want to create a Student(1) or Teacher(2)? [ENTER NUMBER]:'
+    person_input = gets.chomp
+    case person_input
+    when '1'
+      create_student
+    when '2'
+      create_teacher
+    else
+      puts 'Please add a valid input!'
+    end
+    @parent.show_menu
+  end
+
+  def create_teacher
+    print 'Name: '
+    name = gets.chomp
+    print 'Age: '
+    age = gets.chomp
+    print 'Specialization: '
+    specialization = gets.chomp
+    @people.push(Teacher.new(specialization, age, name))
+    puts
+    puts 'Teacher is successfully created'
+    @parent.show_menu
+  end
+
```
- [ ] Good job so far, in your app it is not possible to successfully create a new Student or Person, it would be nice to fix this error  ⬇️ 
![Screenshot_203](https://github.com/stino-x/PST-Classes/assets/69638013/73723597-8348-4c1b-9049-718f22fe9ce8)

prolajumokeoni reviewed (changes requested):
Hi @stino-x 
Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️  Basic UI added



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

stino-x committed:
This commit adds a new method called create_student to the App class.
This method prompts the user for the student's age, name, and whether they have parent permission.
It then creates a new Student object and pushes it onto the @people array.
The method then prints a message indicating that the student was successfully created.

tugsem commented on main.rb:
```diff
@@ -0,0 +1,51 @@
+    puts 'Please choose an option by entering a number:'
+    puts '1 - List all books'
+    puts '2 - List all people'
+    puts '3 - Create a person'
```
- [ ] As requested by the previous reviewer, please fix the option to create a person as these errors occur when I try to create a student or a teacher: 
![image](https://github.com/stino-x/PST-Classes/assets/91486791/7c877265-94e9-489e-8f6d-322418f1fd88)
![image](https://github.com/stino-x/PST-Classes/assets/91486791/ec47a59f-fe10-4500-b625-4e099486264c)

tugsem reviewed (changes requested):
## STATUS: REQUIRED CHANGES ♻️
Hello @stino-x  👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are some issues that you still need to work on to go to the next project but you are almost there!

![ Alt Text](https://media.giphy.com/media/l2Je9zHYveK012EVi/giphy.gif)

 ## Highlights 🚀
✔️Good job on working on the required changes

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

stino-x committed:
The diff adds a new option to the script, which allows users to create a rental.

1GORDON commented on main.rb:
```diff
@@ -0,0 +1,51 @@
+    puts 'Please choose an option by entering a number:'
+    puts '1 - List all books'
+    puts '2 - Create a rental'
+    puts '3 - Create a person'
```
Great job @stino-x 🤝

- [ ] As requested by the previous reviewer, kindly ensure the app is able to create a person. At the moment the app is unable to create the person.
![create student](https://github.com/stino-x/PST-Classes/assets/63917725/f2d4a1cc-5fd7-4b12-9256-fd1b96a10b4f)

1GORDON reviewed (changes requested):
##
## Hi @stino-x ✋

Good job so far!

## Highlights!
- Followed Gitflow ✔️

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

stino-x committed:
This commit adds the ability for teachers to have parent permission and adds a classroom attribute to students.

Felix45 commented on app.rb:
```diff
@@ -0,0 +1,131 @@
+
+  def create_rental
+    puts 'Select a book from the following list by number'
+    @books.each_with_index { |book, index| puts "#{index}) Title: '#{book.title}', Author: #{book.author}" }
+
+    puts 'Enter the number of the book:'
+    book_num = gets.chomp.to_i
+
+    puts 'Select a person from the following list by number (not id)'
+    @people.each_with_index do |person, index|
+      puts "#{index}) [#{person.class}] Name: #{person.name}, ID: #{person.id}, Age: #{person.age}"
+    end
+
+    puts 'Enter the number of the person:'
+    person_num = gets.chomp.to_i
+
+    print 'Date: '
+    date = gets.chomp
+
+    @rentals.push(Rental.new(date, @books[book_num], @people[person_num]))
+    puts 'Rental created successfully'
+    @parent.show_menu
+  end
+
+  def list_of_rentals
+    print 'ID of person: '
+    id = gets.chomp.to_i
+
+    puts 'Rentals:'
+    rentals_found = @rentals.select { |rental| rental.person.id == id }
+    if rentals_found.empty?
+      puts 'No rentals found for the given person ID.'
+    else
+      rentals_found.each do |rental|
+        puts "Date: #{rental.date}, Book \"#{rental.book.title}\" by #{rental.book.author}"
+      end
+    end
+
+    @parent.show_menu
+  end
```
- [ Required ] I think it would be nice if you make sure that the functionality for adding a `Rental` and listing `Rentals` works as expected. Since this functionality is dependent on the `create a person` functionality which was not working before I'll have to mark it as a required change. Please ensure that both `create` and `list` rentals functionality work as expected. I would like to suggest that you refactor your code as shown below for this to work.

## Error when creating a rental
![Screenshot from 2023-10-21 18-30-06](https://github.com/stino-x/PST-Classes/assets/16392046/bb4698dc-fee3-4307-bb44-420573095815)

> I think for these two functionalities to work it would be advisable to refactor the following classes as shown below.

## book.rb

> In the Book class you need to make sure that the `add_rentals` method receives a `date`, and the  `person` renting the book, then create the rental through the `Rental` class constructor as shown below. 

```ruby
require_relative 'rental'

class Book
  attr_accessor :title, :author, :rentals

  def initialize(title, author)
    @title = title
    @author = author
    @rentals = []
  end

  def add_rentals(date, person)
    rental = Rental.new(date, self, person)
    @rentals << rental
  end
end

```

## person.rb
> In the Person class you need to make sure that the `add_rentals` method receives a `date`, and the  `book` being rented out, then create the rental through the `Rental` class constructor as shown below. 

```ruby
require_relative 'nameable' # Include the Nameable module
require_relative 'rental'

class Person < Nameable
  attr_accessor :name, :age, :rentals
  attr_reader :id

  def initialize(name: 'Unknown', age: 0, parent_permission: true)
    super()
    @name = name
    @id = rand(1..1000)
    @age = age
    @parent_permission = parent_permission
    @rentals = []
  end

  private

  def of_age?
    @age >= 18
  end

  def add_rentals(date, book)
    rental = Rental.new(date, book, self)
    @rentals << rental
  end

  public

  def can_use_services?
    of_age? || @parent_permission
  end

  def correct_name
    @name
  end
end

```

stino-x replied:
@Felix45 while i do appreciate i do think its kind of unfair that youd ask me that after three reviwers didnt , the person class and functionality works very well now and i dont think its grounds to overrequest bearung in mind the time constraint

Felix45 replied:
> @Felix45 while i do appreciate i do think its kind of unfair that youd ask me that after three reviwers didnt , the person class and functionality works very well now and i dont think its grounds to overrequest bearung in mind the time constraint

Hi @stino-x, 

I hope you are well, the previous reviewers could not check for the `create_rentals` and `list_rentals` functionality because the `App` was broken. The `create_person` functionality was not working as expected, it's important to note that to create a `rental` you need to select a `person` from a list and this functionality was not working before. Currently, when I run your application and try to create a `Rental` I get the error below. 

## Error when creating a rental
![Screenshot from 2023-10-21 18-30-06](https://github.com/stino-x/PST-Classes/assets/16392046/908078a6-e163-4cd2-b791-05c6a462af37)

I would have loved to approve your project but only if it's working correctly, that's why I gave code suggestions to guide you in implementing this.

Felix45 reviewed (changes requested):
Hi @stino-x, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

<img src="https://camo.githubusercontent.com/59fa9d95c18dbd5a8e47490a9e4b48393f569685410f3c78d4d83a9764fa1392/68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f74385950663341434b656f754c7a454976722f67697068792e676966" width="350px" />

## Highlights :heavy_check_mark: 
- [x] Good console App design :white_check_mark: 
- [x] Git flow is used correctly :white_check_mark: 
- [x] Pull request has a descriptive title and summary :white_check_mark: 
- [x] Implemented all the changes from the previous reviewers :white_check_mark:   

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

stino-x committed:
```
Add Rental class and modify add_rentals method in Book and Person classes.

stino-x committed:
```
Add rental association to Book and Person classes.

This commit adds a rental association to the Book and Person classes.
The Book class now has a rentals attribute that is an array of Rental objects.
The Person class now has a rentals attribute that is an array of Rental objects.
The Rental class has a book attribute that is a Book object and a person attribute that is a Person object.

AmaduKamara reviewed (approved):
# Status: APPROVED 🟢 🟢 🟢

Hi @stino-x, 

Good job working on all highlighted changes requested by previous reviewers 👍🏽

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! 🎉 

## Highlights:

✅ You set up Gitflow correctly 👍🏽
✅ The Linter checks passed successfully 👍🏽
✅ Implemented all features of the UI as expected 👍🏽

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```
